blockstore/splitstore/splitstore.go

```diff
@@ -393,20 +402,46 @@ func (s *SplitStore) AllKeysChan(ctx context.Context) (<-chan cid.Cid, error) {
 		return nil, err
 	}
 
-	ch := make(chan cid.Cid)
+	// arbitrary but >= 1 helps avoid context switching in this case
+	ch := make(chan cid.Cid, 8)
 	go func() {
 		defer cancel()
 		defer close(ch)
 
-		for _, in := range []<-chan cid.Cid{chHot, chCold} {
-			for cid := range in {
-				select {
-				case ch <- cid:
-				case <-ctx.Done():
-					return
+		// uber nit unless we have a reason to do this in order?
+		// Of course, this code is more complicated and also probably overkill.
+		// I actually kind of prefer the old code.
+		// But I'll just leave this here because I already wrote it.
+		for chHot != nil && chCold != nil {
+			var c cid.Cid
+			var ok bool
+			select {
+			case c, ok = <-chHot:
+				if !ok {
+					chHot = nil
+					continue
+				}
+			case c, ok = <-chCold:
+				if !ok {
+					chCold = nil
+					continue
 				}
+			case <-ctx.Done():
+				return
+			}
+
+			select {
+			case ch <- c:
+			case <-ctx.Done():
+				return
 			}
 		}
+
+		// Also note: We're _supposed_ to guarantee that CIDs are only returned once. We
+		// _probably_ need to put the CIDs in a set to check that.
+		//
+		// That might actually be a reason to read through the hotstore before the
+		// coldstore?
```

Comment:
ah yeah, we might have some dupes in the coldstore.
I'll add a markset to make sure we don't return keys twice.